### ISSUE TITLE
Tcollector stops sending data to opentsdb

### ISSUE THREAD

**Issue by nirajanpdl** [OpenTSDB/tcollector#461]
After working for few hours, Tcollector is not sending data to opentsdb below is the log from tcollector.log.
Any suggestion will be highly appreciated, 

2021-11-30 09:01:29,653 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=28038)
2021-11-30 09:01:36,872 tcollector[12201] DEBUG: verifying our TSD connection is alive
2021-11-30 09:01:44,669 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:01:44,672 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=28199)
2021-11-30 09:01:59,688 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:01:59,691 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=28573)
2021-11-30 09:02:14,708 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:02:14,711 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=28929)
2021-11-30 09:02:29,726 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:02:29,729 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=29188)
2021-11-30 09:02:36,876 tcollector[12201] DEBUG: verifying our TSD connection is alive
2021-11-30 09:02:44,740 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:02:44,742 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=29404)
2021-11-30 09:02:59,752 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:02:59,755 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=29793)
2021-11-30 09:03:14,762 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:03:14,765 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=30152)
2021-11-30 09:03:29,782 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:03:29,786 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=30256)
2021-11-30 09:03:36,880 tcollector[12201] DEBUG: verifying our TSD connection is alive
2021-11-30 09:03:44,790 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:03:44,793 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=30616)
2021-11-30 09:03:59,801 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:03:59,804 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=30994)
2021-11-30 09:04:14,819 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:04:14,822 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=31352)
2021-11-30 09:04:29,839 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:04:29,842 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=31463)
2021-11-30 09:04:36,885 tcollector[12201] DEBUG: verifying our TSD connection is alive
2021-11-30 09:04:44,858 tcollector[12201] INFO: cpuUtilisation.py (interval=0) needs to be spawned
2021-11-30 09:04:44,861 tcollector[12201] INFO: spawned cpuUtilisation.py (pid=31816)

**Comment by optiz0r** [comment]
I'm seeing something similar
- tcollector.py version 1.3.3 with some local modifications (listed below)
- VictoriaMetrics instead of OpenTSDB as backend
- Only affects tcollector.py when run under python3 (3.6.8 on EL8 and 3.9.10 on EL9 tested). When running tcollector.py under python2 it operates normally
- strace attached to the metric scripts show they are continuing to output data to stdout right up until they are killed off by tcollector.py
- DEBUG log of tcollector.py shows reads happening normally until everything locks up, then no more reads happen. (i.e. https://github.com/OpenTSDB/tcollector/blob/v1.3.3/tcollector.py#L152 this stops being logged)
- strace of tcollector.py doesn't appear to show any reads happening either. `pselect6` blocks until the timeout is reached and it returns. tcollector then looks for changes in the collector scripts directory, calls wait4() on all the child scripts, and repeats. Occasionally a child script is killed and replaced.

Local modifications:
- Removed the `version` query in `verify_conn` since VictoriaMetrics does not implement any querying functionality in the OpenTSDB compatibility interface. Patch not shown, but it just comments out about half of the verify_conn method, leaving the publication of tcollector's internal metrics.
- Workaround for py3.9 compat; read operation returning None:
```
#in def read(self):
@@ -163,13 +172,18 @@
         # out a bunch of data points at one time and we get some weird sized
         # chunk.  This read call is non-blocking.
         try:
-            self.buffer += self.proc.stdout.read()
+            read = self.proc.stdout.read()
+            if read:
+                self.buffer += read
             if len(self.buffer):
                 LOG.debug('reading %s, buffer now %d bytes',
                           self.name, len(self.buffer))
         except IOError as exc:
             if exc.errno != errno.EAGAIN:
                 raise
+        except TypeError:
+            # https://bugs.python.org/issue13322#msg307763
+            pass
```
- py2/3 compat for str vs bytes
```
@@ -802,7 +818,7 @@
             if self.dryrun:
                 print(out)
             else:
-                self.tsd.sendall(out)
+                self.tsd.sendall(out.encode('utf-8'))
             self.sendq = []
         except socket.error as msg:
             LOG.error('failed to send data: %s', msg)
```

tcollector INFO log once it already hits the bad state:
```
2023-01-13 14:55:41,434 tcollector[1892932] WARNING: Terminating collector nfsstat.py after 603 seconds of inactivity
2023-01-13 14:55:41,436 tcollector[1892932] WARNING: Terminating collector procnettcp.py after 603 seconds of inactivity
2023-01-13 14:55:41,437 tcollector[1892932] WARNING: Terminating collector netstat.py after 603 seconds of inactivity
2023-01-13 14:55:41,438 tcollector[1892932] WARNING: Terminating collector ifstat.py after 603 seconds of inactivity
2023-01-13 14:55:41,439 tcollector[1892932] INFO: Waiting 5s for PID 2494994 (ifstat.py) to exit...
2023-01-13 14:55:42,441 tcollector[1892932] WARNING: Terminating collector softnet.py after 604 seconds of inactivity
2023-01-13 14:55:42,442 tcollector[1892932] WARNING: Terminating collector procstats.py after 604 seconds of inactivity
2023-01-13 14:55:42,444 tcollector[1892932] WARNING: Terminating collector cpustats.py after 604 seconds of inactivity
2023-01-13 14:55:42,445 tcollector[1892932] INFO: Waiting 5s for PID 2494997 (cpustats.py) to exit...
2023-01-13 14:55:43,447 tcollector[1892932] WARNING: Terminating collector buddyinfo.py after 605 seconds of inactivity
2023-01-13 14:55:43,449 tcollector[1892932] INFO: nfsstat.py (interval=0) needs to be spawned
2023-01-13 14:55:43,454 tcollector[1892932] INFO: spawned nfsstat.py (pid=2496157)
2023-01-13 14:55:43,455 tcollector[1892932] INFO: procnettcp.py (interval=0) needs to be spawned
2023-01-13 14:55:43,464 tcollector[1892932] INFO: spawned procnettcp.py (pid=2496158)
2023-01-13 14:55:43,465 tcollector[1892932] INFO: netstat.py (interval=0) needs to be spawned
2023-01-13 14:55:43,476 tcollector[1892932] INFO: spawned netstat.py (pid=2496159)
2023-01-13 14:55:43,476 tcollector[1892932] INFO: ifstat.py (interval=0) needs to be spawned
2023-01-13 14:55:43,486 tcollector[1892932] INFO: spawned ifstat.py (pid=2496160)
2023-01-13 14:55:43,486 tcollector[1892932] INFO: softnet.py (interval=0) needs to be spawned
2023-01-13 14:55:43,494 tcollector[1892932] INFO: spawned softnet.py (pid=2496161)
2023-01-13 14:55:43,495 tcollector[1892932] INFO: procstats.py (interval=0) needs to be spawned
2023-01-13 14:55:43,504 tcollector[1892932] INFO: spawned procstats.py (pid=2496162)
2023-01-13 14:55:43,505 tcollector[1892932] INFO: cpustats.py (interval=0) needs to be spawned
2023-01-13 14:55:43,515 tcollector[1892932] INFO: spawned cpustats.py (pid=2496163)
2023-01-13 14:55:43,516 tcollector[1892932] INFO: buddyinfo.py (interval=0) needs to be spawned
2023-01-13 14:55:43,525 tcollector[1892932] INFO: spawned buddyinfo.py (pid=2496164)
2023-01-13 14:58:43,682 tcollector[1892932] INFO: Heartbeat (12 collectors running)
2023-01-13 15:00:58,823 tcollector[1892932] WARNING: Terminating collector nfsstat_extra.py after 604 seconds of inactivity
2023-01-13 15:00:58,825 tcollector[1892932] INFO: nfsstat_extra.py (interval=0) needs to be spawned
2023-01-13 15:00:58,830 tcollector[1892932] INFO: spawned nfsstat_extra.py (pid=2496640)
2023-01-13 15:04:14,026 tcollector[1892932] WARNING: Terminating collector iostat.py after 605 seconds of inactivity
2023-01-13 15:04:14,028 tcollector[1892932] INFO: iostat.py (interval=0) needs to be spawned
2023-01-13 15:04:14,033 tcollector[1892932] INFO: spawned iostat.py (pid=2497168)
2023-01-13 15:04:29,053 tcollector[1892932] WARNING: Terminating collector dfstat.py after 605 seconds of inactivity
2023-01-13 15:04:29,055 tcollector[1892932] INFO: dfstat.py (interval=0) needs to be spawned
2023-01-13 15:04:29,060 tcollector[1892932] INFO: spawned dfstat.py (pid=2497174)
```

**Comment by rmilkowski** [comment]
The below patch fixes the issue with tcollector stopping collecting stats when run under Python3.

```
commit 0a5a3153febc6fc9c34e4fcf81da37fbff18746f
Author: Milkowski, Robert <Robert.Milkowski@gsacapital.com>
Date:   Fri Feb 3 12:02:22 2023 +0000

    tcollector should not modify dict while iterating over keys() which is a view in Py3.

diff --git a/tcollector.py b/tcollector.py
index 21f9b23..028fbb9 100755
--- a/tcollector.py
+++ b/tcollector.py
@@ -240,7 +240,7 @@ class Collector(object):
           cut_off: A UNIX timestamp.  Any value that's older than this will be
             removed from the cache.
         """
-        for key in self.values.keys():
+        for key in tuple(self.values.keys()):
             time = self.values[key][3]
             if time < cut_off:
                 del self.values[key]
```

**Comment by vasiliyk** [comment]
@optiz0r which python version did you use for PR testing?

**Comment by vasiliyk** [comment]
Closed with PR #456
